### DISCUSSION TITLE
Querying EventHandle Events using API: "message": "Deserialization error, field(message) type is not a EventHandle struct: unexpected end of input"

### DISCUSSION THREAD

**Discussion by kshitijc1506** [aptos-labs/aptos-developer-discussions#99] [Questions] [answered]
Asked by `gftea10086 ` on Discord (#dev-discussion).

> Hi, I publish  the hello_blockchain example to my local net using aptos-node-1.9.7, and try to query the event after set_message call.
> Have read the v1 rest api spec, but it is not working using below syntax
> curl http://127.0.0.1:8080/v1/accounts/0xff89ee9ccec5cf2009c846b2b645af112b581725f57c7d30e92a0d012dd00051/events/0xff89ee9ccec5cf2009c846b2b645af112b581725f57c7d30e92a0d012dd00051::message::MessageChange/from_message
> curl http://127.0.0.1:8080/v1/accounts/0xff89ee9ccec5cf2009c846b2b645af112b581725f57c7d30e92a0d012dd00051/events/0xff89ee9ccec5cf2009c846b2b645af112b581725f57c7d30e92a0d012dd00051::message::MessageHolder/message | jq
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100   159  100   159    0     0   124k      0 --:--:-- --:--:-- --:--:--  155k
> {
>   "message": "Deserialization error, field(message) type is not a EventHandle struct: unexpected end of input",
>   "error_code": "invalid_input",
>   "vm_error_code": null

keywords: API, Events

**Comment by kshitijc1506** [comment]
The error message suggests that the expected type for the message field is not an EventHandle struct. 

The format should be `http://<API_ENDPOINT>/v1/accounts/<ACCOUNT_ADDRESS>/events/<EVENT_KEY>::<MODULE_NAME>::<RESOURCE_NAME>/<EVENT_NAME>`

Can you try running this?
`curl http://127.0.0.1:8080/v1/accounts/0xff89ee9ccec5cf2009c846b2b645af112b581725f57c7d30e92a0d012dd00051/events/0xff89ee9ccec5cf2009c846b2b645af112b581725f57c7d30e92a0d012dd00051::message::MessageHolder/message_change_events`

These steps can be seen on the "[Your First Move Module](https://aptos.dev/tutorials/first-move-module/)" page, under step 5.

**Comment by gftea** [comment]
- the output of resource MessageHolder dose NOT include events
```
curl http://127.0.0.1:8080/v1/accounts/45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc/resourc│
e/0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc::message::MessageHolder | jq                                                                              │
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                │
                                 Dload  Upload   Total   Spent    Left  Speed                                                                                                  │
100   140  100   140    0     0  76211      0 --:--:-- --:--:-- --:--:--  136k                                                                                                 │
{                                                                                                                                                                              │
  "type": "0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc::message::MessageHolder",                                                                        │
  "data": {                                                                                                                                                                    │
    "message": "hello, blockchain"                                                                                                                                             │
  }                                                                                                                                                                            │
}                                                                          

```

- query events returns struct field not found
```
curl http://127.0.0.1:8080/v1/accounts/0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f7│
9db55acfefeadc8fc/events/0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc::message::MessageHolder/message_change_events | jq                                 │
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                │
                                 Dload  Upload   Total   Spent    Left  Speed                                                                                                  │
100   338  100   338    0     0   224k      0 --:--:-- --:--:-- --:--:--  330k                                                                                                 │
{                                                                                                                                                                              │
  "message": "Struct Field not found by Address(0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc), Struct tag(0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9│
f79db55acfefeadc8fc::message::MessageHolder), Field name(message_change_events) and Ledger version(1164)",                                                                     │
  "error_code": "struct_field_not_found",                                                                                                                                      │
  "vm_error_code": null                                                                                                                                                        │
}                                                                                                     


```
- based on move codes, try to query `message` field of MessageHolder
```
curl http://127.0.0.1:8080/v1/accounts/0x45b868338b69eceed4f3dc5ef1dce3fb7c5│
91e10cdda9f79db55acfefeadc8fc/events/0x45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc::message::MessageHolder/message | jq                                   │
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                │
                                 Dload  Upload   Total   Spent    Left  Speed                                                                                                  │
100   159  100   159    0     0  93309      0 --:--:-- --:--:-- --:--:--  155k                                                                                                 │
{                                                                                                                                                                              │
  "message": "Deserialization error, field(message) type is not a EventHandle struct: unexpected end of input",                                                                │
  "error_code": "invalid_input",                                                                                                                                               │
  "vm_error_code": null                                                                                                                                                        │
}                                                                                                      
```

**Reply by kshitijc1506** [reply, marked as answer]
Thanks for the detailed explanation. I think I know what the problem you're experiencing is:

The API calls are trying to query the [`MessageHolder`](https://github.com/aptos-labs/aptos-core/blob/e5349214640cdc3483160cafd3f29c26d5dba319/aptos-move/move-examples/hello_blockchain/sources/hello_blockchain.move#L8) resource for events. However, the resource does not contain any Event Handles. This is likely because ['Event Handles'](https://aptos.dev/concepts/events/#api-access) is deprecated now, and instead we use ['Module Events'](https://aptos.dev/concepts/events/#module-events). While `EventHandle` events still work, developers are encouraged to migrate to Module Events. 

The tutorial over [here](https://aptos.dev/tutorials/first-move-module/) needs to be updated, we will have that done soon. 

**For now, if you'd still like to proceed, you can do this:**
1. Change the module itself- add an Event Handle to the MessageHolder, and in the `set_message` function add the required logic. 
2. Compile and Publish. 
3. Query the event how you were trying to- `http://<API_ENDPOINT>/v1/accounts/<ACCOUNT_ADDRESS>/events/<EVENT_KEY>::<MODULE_NAME>::<RESOURCE_NAME>/<EVENT_NAME>` (the other API calls you were making should also work now)

**Alternatively, you could use Module Events (recommended):** 
The [GraphQL API](https://aptos.dev/guides/system-integrators-guide/#production-network-access) supports module events (and EventHandle events). 

Let me know if this answers your question!

**Reply by gftea** [reply]
The document is indeed misleading, it is using module event, but rest api refers to eventhandle
good to know `The API for querying Module Events is under construction`, I think that explain I can query event `curl http://127.0.0.1:8080/v1/accounts/45b868338b69eceed4f3dc5ef1dce3fb7c591e10cdda9f79db55acfefeadc8fc/events/0x1::coin::CoinStore%3C0x1::aptos_coin::AptosCoin%3E/deposit_events` but cannot query the module event in example code

**Reply by kshitijc1506** [reply]
Yes, that is referring to [`CoinStore`](https://github.com/aptos-labs/aptos-core/blob/e5349214640cdc3483160cafd3f29c26d5dba319/aptos-move/framework/aptos-framework/sources/coin.move#L92), which uses `EventHandle` events.

**Reply by so-schen** [reply]
I also want to know where and how the module event is stored?

**Reply by kshitijc1506** [reply]
[Module events](https://aptos.dev/concepts/events/#module-events) are structs with the [`store`](https://aptos.dev/move/book/abilities/#store) and `drop` abilities. Therefore, they are stored in [global storage](https://aptos.dev/move/book/global-storage-structure/) as a part of the module's bytecode. However, you can also store them as a part of a top-level resource in global storage since you can include them as a field for a resource (since module events have the `store` ability). 

For additional reference, here's an [example](https://explorer.aptoslabs.com/account/0x66c34778730acbb120cefa57a3d98fd21e0c8b3a51e9baee530088b2e444e94c/modules/code/event?network=testnet) of how to use Module Events.

**Reply by gftea** [reply]
got it, thanks!
